mesonbuild/coredata.py

```diff
@@ -349,6 +349,51 @@ def set_builtin_option(self, optname, value):
             raise RuntimeError('Tried to set unknown builtin option %s.' % optname)
         self.builtins[optname].set_value(value)
 
+        # Make sure that buildtype matches other settings.
+        if optname == 'buildtype':
+            self.set_others_from_buildtype(value)
+        else:
+            self.set_buildtype_from_others()
+
+    def set_others_from_buildtype(self, value):
+        if value == 'plain':
+            opt = '0'
```

Comment:
Yes. However it would mean adding an "empty" value for the optimization level which seems a bit redundant. How about if we change the value `0` to mean no flags, rather than adding `-O0`?

Reply:
Hmm, but then that adds an off-by-one in our optimization levels. `opt_level=1` is `-O0` and so on.

Reply:
I think we could solve this by setting `opt = None` and then below not call `self.builtins['optimization'].set_value(opt)` at all if opt is None.

Reply:
No, an off by one would be stupid. In the proposed approach value 0 would add no optimization libs at all (even -O0). This would mirror warning level, where a level of zero is no warning flags at all. AFAIK all compilers default to no optimization if no args are specified and that `-O0` is the same as having no optimization flags.

The important question here is is there a need to have a setting that _explicitly_ adds `-O0`. I can't think of one at least.